Add executable to extract and process ZIP+4 details, grouping zip codes with county codes

tamaradement opened the pull request:
## PR Description
Adds an executable that utilizes the zip4 package to extract zip4Details from a valid tar file, groups five-digit ZIP codes with all corresponding county codes, and outputs to the console.

For context: this is the first PR that completes the first step in an implementation that will eventually do all of the following:
- [X] Write a program that leverages the zip4 package to print all counties associated with a zip code.
- [ ] Output data to a CSV (Note: currently, we're outputting to the console).
- [ ] Implement a function to convert county codes to county names. This will make data analysis easier.
- [ ] Use the output or a similar program to find differences between USPS and the [Smarty US ZIP Code API](https://www.smarty.com/docs/cloud/us-zipcode-api).


Test Plan:
- Ran the program with a valid tar file and verified correct output.
- Checked error messages for scenarios with a missing tar file argument.
- Checked error messages for scenarios with an invalid tar file.

Sample Output:
```
// ... more entries ...
ZIP Code: 53109, County Numbers: [059]
ZIP Code: 81084, County Numbers: [009 099]
ZIP Code: 50940, County Numbers: [153]
ZIP Code: 56167, County Numbers: [105 063]
```

## PR Checklist
* [no] **New automated tests have been written to the extent possible.**
* [yes, ran `go build` without incident ] **The code has been checked for structural/syntactic validity.**
	- AMI/application: a build was performed
	- terraform changes: "terraform plan" checked on every affected environment
* [n/a ] **(If applicable) the code has been manually tested on our infrastructure.**
	- AMI/application: deployed an a test or dev environment
	- terraform changes: applied to test or dev environment
	- script: run against test or dev environment
* [n/a] **Likely failure points and new functionality have been identified and tested manually.**
	Examples:
	- Application manually run in a way that triggers any new branches
	- AMI logged into and changes verified from login shell
* [yes, see above] **Pull request description includes a description of all the manual steps performed to accomplish the above.**

To provide feedback on this template, visit https://docs.google.com/document/d/1YfTv7Amyop5G_8w1c2GJ_Mu-70L0KkZHhm9f9umDi3U/edit

Commit:
Adds initial implementation for ZIP+4 Query Tool

Adds an executable that utilizes the zip4 package to extract zip4Details from a valid tar file; groups five digit zips with all corresponding county codes and outputs to the console.

Tested:
- Ran the program with a valid tar file and observed the output.
- Verified error messages for missing tar file argument.
- Verified error messages for non-existent tar file.

Commit:
Improve error handling and function naming for clarity

This commit updates the error handling to use fmt.Fprintf + os.Exit(1) pattern for better control over output. Additionally, it renames the function CollectDetails to the more descriptive CollectZip4Details.

Testing:
- Ran the program with a valid tar file and verified correct output.
- Checked error messages for scenarios with a missing tar file argument.
- Checked error messages for scenarios with an invalid tar file.

jonahb commented on counties/counties.go:
```diff
@@ -0,0 +1,50 @@
+package counties
```
Are you thinking these functions will be used from multiple commands?

If this package supports stuff in this module, good to put it under `/internal`. Here's a post that talks about internal packages: https://go.dev/doc/go1.4#internalpackages.

Also, it's great that you submitted the PR to this repo to avoid blocking on EUA, etc. In the end, this program might fit in a private repo we have called `zipreports`.

tamaradement replied:
Ok cool, good to know! I haven't identified a clear use case for this package outside of this module, so I'll take your advice and move it into `/internal` for now.

jonahb commented on counties/counties.go:
```diff
@@ -0,0 +1,50 @@
+func CollectZip4Details(tarName, zipPassword string) (map[string][]string, error) {
+	var details []zip4.Zip4Detail
+
+	yield := func(detail zip4.Zip4Detail) {
```
Reading everything into `details` may use a lot of memory and slow things down. It would be interesting to compare the results without the copy. In other words, what if `yield` contained the logic from `mapZipToCounty`?

tamaradement replied:
Makes sense! I've refactored the yield function to incorporate the logic from `mapZipToCounty`.

jonahb replied:
Cool, were you able to see any difference in running time or memory usage?

tamaradement replied:
The response time does seem snappier, although I haven't formally timed it. Regarding space, by eliminating the in-between step of storing all the details, we should theoretically reduce the memory footprint from 2n to n, where n = (the number of zip4Details being processed). Am I understanding this correctly?

jonahb replied:
I think theoretically the memory usage in this part of the program could go from linear (in the number of records) to constant. `ReadCityStateFromZip4Tar` allocates memory for each record. If they're all kept in a slice, that'll take memory proportional to the number of records. If they're processed one-by-one, you only need one in memory at a time. Go can deallocate (garbage collect) memory whenever it likes, so I don't think the real difference would be this dramatic, but I'd expect to see something!

jonahb commented on counties/counties.go:
```diff
@@ -0,0 +1,50 @@
+		county := detail.CountyNumber
+
+		// Check if the county number is already in the list for the zip.
+		counties, exists := zipToCounty[zip]
```
`zipToCounty[zip]` returns a nil slice (the zero value of `[]string`) if `zip` doesn't exist in the map, and nil slices behave like any zero-length slice. So I don't think you need the `exists` here or the `if` below.

tamaradement replied:
Ah, gotcha. I'm still getting used some of go's nuances. I've updated the logic to eliminate the redundant check. When time allows, please take a look and let me know what you think of the updated organization.

jonahb reviewed (commented):
This is a great start. I like how you outlined the plan in the PR description. A few comments to think about.

Commit:
Optimize memory usage of counties logic

Refactored the logic within the yield function to directly update zip-to-county mappings as data is read.
Additionally, moved the 'counties' package to the 'internal' directory to restrict its accessibility outside the module.
Removed the redundant existence check for zip codes in the map,

Tested:
    - Ran the program with a valid tar file and observed output.
    - Checked error messages for scenarios with a missing tar file argument.
    - Checked error messages for scenarios with an invalid tar file.

jonahb reviewed (approved):
Looks great!
